### PR TITLE
Make VPA maxAllowed configurable in the Seed and Garden resource specs

### PR DESCRIPTION
**Pull request by ialidzhikov**
**How to categorize this PR?**

/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
This PR is a rework of https://github.com/gardener/gardener/pull/10208.

For components running in the Seed cluster, we would like to prevent VPA from recommending more than the largest Nodes allocatable. This is a [know limitation](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/README.md#known-limitations) of VPA.

This PR is a first step towards the final goal. For the final goal we would like to have all VPAs covered. This PR is currently covering the components at highest risk: etcd, apiservers and prometheus.

As a next step, we would like to evaluate different approaches for covering all VPAs:
- Leverage a webhook in the Seed that mutates maxAllowed of all VPAs 
- [vpa-recommender: Make max allowed recommendation configurable](https://github.com/kubernetes/autoscaler/issues/7147)
- Again upstream feature request, implement in the upstream what GKE have: vpa-updater and vpa-admission-controller talking to the cluster-autoscaler via the ProvisionnngRequest API of cluster-autoscaler.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
N/A

**Release note**:

```feature operator
The VerticalPodAutoscaler's maximum allowed recommendation is now configurable in the Seed and Garden resource. The reasoning for this feature is a [know limitation](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/README.md#known-limitations) where VPA is unaware of the largest Nodes' allocatable resources and can wrongly recommend resource requests which make the Pod unschedulable. To mitigate this issue the Seed and Garden resources now support configured VPA maxAllowed. Currently the effect of this field is only limited to the components at highest risk such as etcd, kube-apiserver and prometheus. In future, we would like to cover all VPAs in the Seed and garden runtime cluster to have a configurable maxAllowed that we recommend to be the largest cluster Node's allocatable resources (minus the DeamonSet Pods resource requests).
```

### CODE REVIEW

**Comment by ialidzhikov** [comment]
`pull-gardener-unit` fails with:

```
2024/08/28 15:19:25 Failed executing local generator: some packages had errors:
type github.com/gardener/gardener/pkg/apis/core/v1beta1.NginxIngress cannot be converted to protobuf: unable to embed type "*k8s.io/api/core/v1.ServiceExternalTrafficPolicyType" as field "externalTrafficPolicy" in "github.com/gardener/gardener/pkg/apis/core/v1beta1.NginxIngress": did not recognize the provided type
make[1]: *** [/home/prow/go/src/github.com/gardener/gardener/Makefile:198: generate] Error 1
make[1]: Leaving directory '/home/prow/go/src/github.com/gardener/gardener'
make: *** [Makefile:142: check-generate] Error 1
```


Locally `make check-generate` works for me:
```
% make check-generate
> Generate
>> make generate
>> make tidy
```

🤔

**Comment by ialidzhikov** [comment]
Ok, it is not failing only for this PR. Same failure in https://github.com/gardener/gardener/pull/10400#issuecomment-2316859765

@oliver-goetz can it be related to go1.23 or any other recent change?

**Comment by ialidzhikov** [comment]
cc @voelzmo @plkokanov @vlerenc

**Comment by oliver-goetz** [comment]
> Ok, it is not failing only for this PR. Same failure in [#10400 (comment)](https://github.com/gardener/gardener/pull/10400#issuecomment-2316859765)
> 
> @oliver-goetz can it be related to go1.23 or any other recent change?

It seems to be related to a new version of Protocol Buffers v27.4. The maintainers recreated the release yesterday (see https://github.com/gardener/gardener/pull/10416).

**Comment by ialidzhikov** [comment]
```
{"component":"entrypoint","file":"sigs.k8s.io/prow/pkg/entrypoint/run.go:264","func":"sigs.k8s.io/prow/pkg/entrypoint.gracefullyTerminate","level":"error","msg":"Process gracefully exited before 10m0s grace period","severity":"error","time":"2024-08-29T13:33:28Z"}
```

**Comment by ialidzhikov** [comment]
I rebased to latest master.

File: docs/operations/seed_settings.md [hunk 1]
```diff
@@ -122,6 +122,15 @@ By setting the `.spec.settings.verticalPodAutoscaler.enabled=false`, you can dis
 
 ⚠️ In any case, there must be a VPA available for your seed cluster. Using a seed without VPA is not supported.
 
+### Vertical Pod Autoscaler Maximum Allowed Recommendation
+
+One of the [known limitations](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/README.md#known-limitations) of VPA is that it is unaware of the Node's capacity, and can increase the resource requests of a Pod beyond the capacity of any single Node. Such Pod is likely to become permanently unschedulable. That problem can be partly mitigated by using the VerticalPodAutoscaler's `.spec.resourcePolicy.containerPolicies[].maxAllowed` field to constrain Pod resource requests to the level of Nodes' allocatable resources to make sure that the Pod will be always schedulable.
```
Comment (rfranzke) [inline docs/operations/seed_settings.md:127]: Please put each sentence into their own line. This simplifies PR reviews

File: docs/operations/seed_settings.md [hunk 2]
```diff
@@ -122,6 +122,15 @@ By setting the `.spec.settings.verticalPodAutoscaler.enabled=false`, you can dis
...
+
+By setting the `.spec.settings.verticalPodAutoscaler.maxAllowed` field to the largest Seed Nodes' allocatable resources you can make sure that no Pods will be unschedulable due to the above-mentioned VPA limitation.
```
Comment (rfranzke) [inline docs/operations/seed_settings.md:129]:
```suggestion
By setting the `.spec.settings.verticalPodAutoscaler.maxAllowed` field to the largest Seed Nodes' allocatable resources, you can make sure that no Pods will become unschedulable due to the above-mentioned VPA limitation.
```

File: docs/operations/seed_settings.md [hunk 3]
```diff
@@ -122,6 +122,15 @@ By setting the `.spec.settings.verticalPodAutoscaler.enabled=false`, you can dis
...
+
+> [!WARNING]  
+> Currently, the effect of setting this field is only limited to VPAs of the following components: etcd, kube-apiserver and prometheus.
```
Comment (rfranzke) [inline docs/operations/seed_settings.md:132]: Which Prometheus? There are multiple running in a seed. All?

**Review by rfranzke — CHANGES_REQUESTED** [review]
I am wondering about the approach of "hard-coding" this into each and every component. Currently, it's only done for `etcd`, `kube-apiserver`, and `prometheus`, but I guess we would like this have all components running in the seed configured with this  `maxAllowed` setting, right? This is at least what the TODO statement suggests. In this light, wouldn't it make more sense to introduce a webhook in `gardener-resource-manager` that handles this across the components in the seed? This way, we could easily handle components deployed via extensions as well.

**Comment by ialidzhikov** [comment]
> I am wondering about the approach of "hard-coding" this into each and every component. Currently, it's only done for `etcd`, `kube-apiserver`, and `prometheus`, but I guess we would like this have all components running in the seed configured with this `maxAllowed` setting, right? This is at least what the TODO statement suggests. In this light, wouldn't it make more sense to introduce a webhook in `gardener-resource-manager` that handles this across the components in the seed? This way, we could easily handle components deployed via extensions as well.

The implementation plan was already prepared. In this PR I simply reworker/reopened https://github.com/gardener/gardener/pull/10208. I think we were "in a hurry" for this feature as it was causing issues on our side. I am not sure if it is still the case after we have the worker pools with the larger machine sizes.
For the GRM webhook approach or the upstream feature request (https://github.com/kubernetes/autoscaler/issues/7147) - I didn't look into all details but the problem there is rather theoretical: we want the maxAllowed value to be per Pod and for Pods with many containers how to distribute the maxAllowed values.

**Comment by rfranzke** [comment]
OK, but what now? Shall we close this PR (it doesn't seem to be of much interest given its staleness) and wait for the upstream improvement? If you want to continue with the PR, I still vote for a more generic approach as explained above.

**Pull request closed without merging**